File: proyectosDAW/proxectoPHP_Produtos/models/ProdutoDAO.php
<?php

class ProdutoDAO implements DAO
{

    public function getAll()
    {
        $produtos = [];
        $conexion = ConexionBD::obterConexion();

        try {

            $sql = "SELECT * FROM produto ORDER BY id ASC";

            $stmt = $conexion->prepare($sql);
            $stmt->execute();
            $results = $stmt->fetchAll(PDO::FETCH_OBJ);

            foreach ($results as $r) {
                $res = new Produto($r->nome, $r->prezo, $r->descricion, $r->activado, $r->idCategoriaProduto);
                $res->setId($r->id);
                $produtos[] = $res;
            }
        } catch (PDOException $e) {
            echo "Houbo problemas obtendo a listaxe de produtos. ProdutoDAO getAll " . $e->getMessage();
        }

        return $produtos;
    }

    public function get($id)
    {
        $produto = null;
        $conexion = ConexionBD::obterConexion();

        try {

            $sql = "SELECT * FROM produto WHERE id LIKE ? ";

            $stmt = $conexion->prepare($sql);
            $stmt->bindValue(1, $id);
            $stmt->execute();
            $results = $stmt->fetchAll(PDO::FETCH_OBJ);
            if ($stmt->rowCount() > 0) {
                $r = current($results);
                $produto = new Produto($r->nome, $r->prezo, $r->descricion, $r->activado, $r->idCategoriaProduto);
                $produto->setId($r->id);
            }
        } catch (PDOException $e) {
            echo "Houbo problemas obtendo o produto. ProdutoDAO get " . $e->getMessage();
        }

        return $produto;
    }

    public function save($object)
    {
        $conexion = ConexionBD::obterConexion();

        try {
            $sql = "SELECT id FROM produto WHERE nome LIKE ? ";
            $stmt = $conexion->prepare($sql);
            $stmt->bindValue(1, $object->getNome());
            $stmt->execute();
            if ($stmt->rowCount() == 0) {
                $sql = "INSERT INTO produto (nome, prezo, descricion, activado, idCategoriaProduto) VALUES (?, ?, ?, ?, ?) ";

                $stmt = $conexion->prepare($sql);
                $stmt->bindValue(1, $object->getNome());
                $stmt->bindValue(2, $object->getPrezo());
                $stmt->bindValue(3, $object->getDescricion() != NULL ? $object->getDescricion() : NULL);
                $stmt->bindValue(4, $object->getActivado());
                $stmt->bindValue(5, $object->getIdCategoriaProduto());

                $stmt->execute();
            } else {
                echo "O produto introducido xa existe";
            }
        } catch (PDOException $e) {
            echo "Houbo problemas gardando o produto. ProdutoDAO save " . $e->getMessage();
            return false;
        }

        return true;
    }

    public function update($object)
    {
        $conexion = ConexionBD::obterConexion();

        try {
            $sql = "UPDATE produto SET nome = ?, prezo = ?, descricion = ?, activado = ?, idCategoriaProduto = ? WHERE id LIKE ?";

            $stmt = $conexion->prepare($sql);
            $stmt->bindValue(1, $object->getNome());
            $stmt->bindValue(2, $object->getPrezo());
            $stmt->bindValue(3, $object->getDescricion() != NULL ? $object->getDescricion() : NULL);
            $stmt->bindValue(4, $object->getActivado());
            $stmt->bindValue(5, $object->getIdCategoriaProduto());
            $stmt->bindValue(6, $object->getId());
            $stmt->execute();
        } catch (PDOException $e) {
            echo "Houbo problemas actualizando o produto. ProdutoDAO update" . $e->getMessage();
            return false;
        }

        return true;
    }

    public function delete($object)
    {
        $conexion = ConexionBD::obterConexion();

        try {
            $sql = "DELETE FROM produto WHERE id LIKE ?";

            $stmt = $conexion->prepare($sql);
            $stmt->bindValue(1, $object->getId());
            $stmt->execute();
        } catch (PDOException $e) {
            echo "Houbo problemas eliminando o produto. ProdutoDAO delete " . $e->getMessage();
            return false;
        }
    }
}
?>

File: proyectosDAW/MVC_Servlet_final/src/com/acarballeira/model/ProdutoDAO.java
package com.acarballeira.model;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;

import com.acarballeira.ConexionBD;

public class ProdutoDAO implements DAO<Produto> {

	@Override
	public Produto get(int id) {
		Connection c = null;
		PreparedStatement st = null;
		ResultSet rs = null;
		Produto p = null;
		
		try {
			String sql = "SELECT id, nome, prezo, descricion, stockMinimo, stockActual " 
					+ "FROM produto "
					+ "WHERE id = ? "
					+ "ORDER BY nome, prezo";
			c = ConexionBD.obterConexion();
			st = c.prepareStatement(sql);
			st.setInt(1, id);
			rs = st.executeQuery();
			
			if (rs.next()) {
				p = new Produto(rs.getInt(1), rs.getString(2), rs.getDouble(3), rs.getString(4), rs.getInt(5), rs.getInt("stockActual"));
			}
				
			rs.close();
			st.close();
			
		} catch(SQLException e) {
			System.out.println("Erro en ProdutoDAO - get: " + e.getMessage());
		} finally {
			ConexionBD.devolverConexion(c);
		}
		
		return p;
	}

	@Override
	public ArrayList<Produto> getAll() {
		Connection c = null;
		PreparedStatement st = null;
		ResultSet rs = null;
		ArrayList<Produto> listaP = new ArrayList<Produto>();
		
		try {
			String sql = "SELECT id, nome, prezo, descricion, stockMinimo, stockActual " 
					+ "FROM produto "
					+ "ORDER BY nome, prezo";
			c = ConexionBD.obterConexion();
			st = c.prepareStatement(sql);
			rs = st.executeQuery();
			
			while(rs.next()) {
				Produto p = new Produto(rs.getInt(1), rs.getString(2), rs.getDouble(3), rs.getString(4), rs.getInt(5), rs.getInt("stockActual"));
				listaP.add(p);
			}
				
			rs.close();
			st.close();
			
		} catch(SQLException e) {
			System.out.println("Erro en ProdutoDAO - getALL: " + e.getMessage());
		} finally {
			ConexionBD.devolverConexion(c);
		}
		
		return listaP;
	}

	@Override
	public boolean save(Produto p) {
		boolean rexistrar = false;
		Connection c = null;
		PreparedStatement st = null;
		
		try {
			String sql = "INSERT INTO produto (nome, prezo, descricion, stockMinimo, stockActual) " 
					+ "VALUES (?, ?, ?, ?, ?)";
			c = ConexionBD.obterConexion();
			st = c.prepareStatement(sql);
			st.setString(1, p.getNome());
			st.setDouble(2, p.getPrezo());
			st.setString(3, p.getDescricion());
			st.setInt(4, p.getStockActual());
			st.setInt(5, p.getStockMinimo());
			st.executeUpdate();
			rexistrar = true;	
			st.close();
			
		} catch(SQLException e) {
			System.out.println("Erro en ProdutoDAO - save: " + e.getMessage());
		} finally {
			ConexionBD.devolverConexion(c);
		}
		
		return rexistrar;
	}

	@Override
	public boolean update(Produto p) {
		boolean rexistrar = false;
		Connection c = null;
		PreparedStatement st = null;
		
		try {
			String sql = "UPDATE produto SET nome = ?, prezo = ?, descricion = ?, stockMinimo = ?, stockActual = ? " 
					+ "WHERE id = ?";
			c = ConexionBD.obterConexion();
			st = c.prepareStatement(sql);
			st.setString(1, p.getNome());
			st.setDouble(2, p.getPrezo());
			st.setString(3, p.getDescricion());
			st.setInt(4, p.getStockActual());
			st.setInt(5, p.getStockMinimo());
			st.setInt(6, p.getId());
			st.executeUpdate();
			rexistrar = true;	
			st.close();
			
		} catch(SQLException e) {
			System.out.println("Erro en ProdutoDAO - update: " + e.getMessage());
		} finally {
			ConexionBD.devolverConexion(c);
		}
		
		return rexistrar;
	}

	@Override
	public boolean delete(Produto p) {
		boolean rexistrar = false;
		Connection c = null;
		PreparedStatement st = null;
		
		try {
			String sql = "DELETE FROM produto " 
					+ "WHERE id = ?";
			c = ConexionBD.obterConexion();
			st = c.prepareStatement(sql);
			st.setInt(1, p.getId());
			st.executeUpdate();
			rexistrar = true;	
			st.close();
			
		} catch(SQLException e) {
			System.out.println("Erro en ProdutoDAO - delete: " + e.getMessage());
		} finally {
			ConexionBD.devolverConexion(c);
		}
		
		return rexistrar;
	}

}
